specification/dartLangSpec.tex

```diff
@@ -16244,10 +16245,19 @@ \subsection{Identifier Reference}
   where $T$ is a subtype of any other type $S$
   such that $v$ is known to have type $S$,
   in which case the static type of $e$ is $T$.
+\item
+  % Extension getter invocation by scope.
+  If $D$ is a declaration of an instance getter in an extension $E$
+  then the static type of $e$ is the return type of $D$.
+\item
+  % Extension method tear-off by scope.
+  If $D$ is a declaration of an instance method
+  in an extension $E$ then the static type of $e$ is
+  the function type of $D$.
 \item
   \commentary{%
     A lexical lookup will never yield a declaration $D$
-    which is an instance member.%
+    which is an instance member of a class.%
```

Comment:
Why not? (Github is preventing me from seeing most of the context here, but this is "Unqualified Invocation", right?)

Do we assume that it's *already* rewritten to `this.id(...)`? (If so, say so).


Also, while we are here, above (where Github won't let me see) it says:
```latex
\item
  When $D$ is a type declaration, that is,
  a declaration of a class, mixin, type alias, or type parameter,
  the following applies:
```
It needs to also mention `enum` (unless that's implicitly declaring a class, which it probably, technically, is. but it's confusing when a "declaration of a class" and a "class declaration" is not the same thing).

Reply:
The enclosing section is still 'Identifier Reference', not 'Unqualified Invocation'.

The reason why lexical lookup won't ever yield a declaration that is an instance member of a class is that it yields a member signature in that case.

I did it like that because this ensures that we have a simple rule for when to add `\THIS{}` in front of an expression (identifier reference or unqualified invocation), and that's exactly when the result from the lexical lookup is a member signature.

(That member signature is actually never used, it's just taken as a signal that we should prepend `\THIS.` to the expression and proceed with that; so I've changed that in another PR, reviewed by Leaf, such that we just yield a "special value" indicating that we should prepend `\THIS.`).

> Do we assume that it's already rewritten to `this.id(...)`?

No, lexical lookup has selected a declaration $D$, and that's never a member of a class interface, because (both for an inherited member and for a locally declared member) we always obtain a member signature in that case. That's the case approx. 10 lines further down from here.

Reply:
Added 'enumerated type' to the list of kinds of type declarations that we may encounter.

Reply:
So we did a hack in "lexical lookup" to not actually find "a declaration", even when there literally is an instance declaration with that name in the surrounding class. Might be worth reminding the reader about :)